File: leetcode/util_functions/java/src/TreeNodeUtils.java
/**
 * Umpire Repl.it Utility Functions Java Edition
 * TreeNode functions are in here.
 *
 * Macarthur Inbody admin-contact@transcendental.us
 * License: CC0
 *
 * Basically just copy the methods from this file into your repl to make it easier
 * to work with. The code is bad I know that but it works so it's better than nothing.
 */

//you'll have to include these 2 methods in your imports.
import java.util.ArrayDeque;
import java.util.Deque;
//this isn't needed unless you want to use an ArrayList sort of like Python's list class.
import java.util.ArrayList;
//this is only used for the main method here but if you just copy the methods you won't need it.
import java.util.Arrays;


public class TreeNodeUtils {
	public static class TreeNode{
		Integer val;
		TreeNode left = null;
		TreeNode right = null;
		TreeNode(){}
		TreeNode(Integer val){ this.val = val;}
	}
	public static void insertNode(TreeNode node, Integer key){
		if(node == null){
			node = new TreeNode(key);
			return;
		}
		Deque<TreeNode> q = new ArrayDeque<>();
		q.add(node);
		while(!q.isEmpty()){
			node = q.removeFirst();
			if(node.left == null){
				node.left = new TreeNode(key);
				break;
			}
			else
				q.add(node.left);
			if(node.right == null){
				node.right = new TreeNode(key);
				break;
			}
			else
				q.add(node.right);
		}
	}

	public static TreeNode arrayToTreeNode(Integer[] arr){
		if(arr.length == 0)
			return null;
			
		int arr_len = arr.length;
		class x_ {
			TreeNode _rec_insert(TreeNode root, Integer i) {
				if(i < arr_len){
					root = new TreeNode(arr[i]);
					int l = (i << 1) + 1;
					root.left = _rec_insert(root.left, l);
					root.right = _rec_insert(root.right, l + 1);
				}
				return root;
			}
		}
		TreeNode root = null;
		root = new x_()._rec_insert(root,0);
		return root;
	}

	public static Integer[] treeNodeToArray(TreeNode root,Integer len){
		Deque<TreeNode> q = new ArrayDeque<>();
		q.add(root);
		Integer[] data = new Integer[len];
		int i = 0;
		while(!q.isEmpty()){
			TreeNode cur = q.removeFirst();
			data[i++] = cur.val;
			if(cur.left != null)
				q.add(cur.left);
			if(cur.right != null)
				q.add(cur.right);

		}
		return data;
	}

	public static ArrayList<Integer> treeNodeToArrayList(TreeNode root){
		Deque<TreeNode> q = new ArrayDeque<>();
		q.add(root);
		ArrayList<Integer> data = new ArrayList<>();
		while(!q.isEmpty()){
			TreeNode cur = q.removeFirst();
			data.add(cur.val);
			if(cur.left != null)
				q.add(cur.left);
			if(cur.right != null)
				q.add(cur.right);
		}
		return data;
	}
	public static void main(String[] args){
		Integer[] arr = {1,2,2,3,3,null,null,4,4};//{1,2,3,4,5,6};
		TreeNode t = arrayToTreeNode(arr);
		Integer[] out = treeNodeToArray(t,arr.length);
		ArrayList<Integer> out2 = treeNodeToArrayList(t);

		System.out.println(Arrays.toString(arr));
		System.out.println(Arrays.toString(out));
		System.out.println(out2);
	}
}

File: leetcode/util_functions/python/treenode_utils.py
from queue import deque
from typing import Optional

"""
LeetCode TreeNode util functions to make your repls/on your own code easier.
By: Macarthur Inbody
License: CC0

Note that this file uses the standard 4 spaces per indent instead of tabs(as the rest of the repo does).


Example:
	Say the function is invert a binary tree.
	You have a function called invert_tree
	
	You'd use the function below like so.
	
	First we input the binary tree list.
	test = [4,2,7,1,3,6,9]
	
	Then we'd convert it to a binary tree.
	binary_tree = list_to_bt(test)
	
	Then you'd be able to pass it to the function like normal!
	This way you can use your own code and especially during repls.
	result = invert_tree(binary_tree)
	
"""


class TreeNode:
	def __init__(self, val=0, left=None, right=None):
		self.val = val
		self.left = left
		self.right = right

	def __repr__(self):
		return f'{self.val} l=>{self.left} r=>{self.right}'


def insert_node(node: TreeNode, key: int) -> None:
	"""
	Insert a TreeNode into the Binary Tree

	:param node: TreeNode object representing root.
	:param key: The value to insert.
	:return: None
	"""
	if not node:
		node = TreeNode(key)
		return
	q = deque((node))
	while len(q):
		node = q.popleft()
		if not node.left:
			node.left = TreeNode(key)
			break
		else:
			q.append(node.left)
		if not node.right:
			node.right = TreeNode(key)
			break
		else:
			q.append(node.right)


def list_to_bt(test_list: list) -> Optional[TreeNode]:
	"""
	Convert the input list into a Binary Tree.

	:param test_list: The list of values we need to use to create our Binary Tree.
	:return: The created Binary Tree or None if the list is empty.
	"""

	def _rec_insert(root,i):
		if i < test_len:
			root = TreeNode(test_list[i])
			l = (i<<1)+1
			root.left = _rec_insert(root.left,l)
			root.right = _rec_insert(root.right,l+1)
		return root
	if len(test_list) == 0:
		return None
	test_len = len(test_list)
	root = None
	root = _rec_insert(root,0)
	return root


def bt_to_list(root: TreeNode) -> list[int]:
	"""
	Converts a Binary Tree into a Linked List.

	:param root: The root node
	:return: A list of integers representing the tree in bfs organization.
	"""

	# use a deque b/c we're always popping from front.
	queue = deque([root])
	# our data list
	data = []
	# while we have data in it.
	while len(queue):
		# make the current element be the first element.
		cur = queue.popleft()
		# if it's None just insert None into it.
		if cur is None:
			data.append(None)
		else:
			# insert the value.
			data.append(cur.val)
			# if we have a left leaf add it to the q.
			if cur.left: queue.append(cur.left)
			# same with the right.
			if cur.right: queue.append(cur.right)
	# then we return the data.
	return data


bt = list_to_bt(['a','b','c','d','e','f'])
print(bt_to_list(bt))

